File: LC/[1959] Minimum Total Space Wasted With K Resizing Operations.cpp
class Solution {
public:
    int minSpaceWastedKResizing(vector<int>& nums, int k) {
        int n = nums.size();//all elements need to fit in the dynmic array
        /*
        dp[ii][jj]: min space wasted on iith element, with jj resizing
                          sz
        x x x x x [x x ii]

        for s in [0, ii+1]:
            dp[ii][jj] = min(dp[ii][jj], dp[s-1][jj-1] + WasteSpace(S:ii+1))
            where WasteSpace(s:ii+1) = max[s:ii+1]*(ii-s+1) - intervalSum[s:ii+1]
        */
        int dp[n][k+1];//declare the dp with size of input array, k+1 resizing capacity

        //initialize the array to max (we are looking for min wasted space)
        for(int ii=0; ii<n; ++ii){
            for(int jj=0; jj<=k; ++jj){
                dp[ii][jj]=INT_MAX/2;
            }
        }

        int mx=0;
        int sum=0;
        for(int ii=0; ii<n; ++ii){//initialize when jj=0. (there is no jj-1)
            mx = max(mx, nums[ii]);
            sum+=nums[ii];
            dp[ii][0]=mx*(ii+1)-sum;
        }

        for(int ii=1; ii<n; ++ii){//for each element.  check where could idx out of range
            for(int jj=1; jj<=min(ii, k); ++jj){//try all the resizing
              
                int mx=0;
                int intervalSum=0;
                //update dp[ii][jj]
                for(int ss=ii; ss>=1; --ss){//for each ii as the ending element, try all to be in this resizing and capture the min
                    mx=max(mx, nums[ss]);
                    intervalSum += nums[ss];
                    dp[ii][jj] = min(dp[ii][jj], dp[ss-1][jj-1]+mx*(ii-ss+1) - intervalSum);
                }
            }
        }

        int ans = INT_MAX/2;
        for(int jj=0; jj<=k; ++jj){
            ans = min(ans, dp[n-1][jj]);
        }
        return ans;
    }
};

File: LC/[1959] Minimum Total Space Wasted With K Resizing Operations.java
class Solution {
    public int minSpaceWastedKResizing(int[] nums, int k) {
        //min waste given an array and cost formula
        //for each ii, try all the valid size (color)
        //1473 color is given, 1959, valid size is all the max upto ii <<<<< 
        //1473 try all the colors -> this is a type I DP
        //1959 try all valid size -> this is an interval DP. last interval on top of prior interval

        int N = nums.length;

        //declare/define DP
        //1473 we use 3D dp cuz the number of color is given
        //1959, we do not know how many resize (actually, could be N)
        int[][] dp = new int[N][k+1]; //dp[ii][jj]: min waste upto ii with jj resize

        //initialize dp
        //default to MAX as we are minimize the waste
        for(int ii=0; ii<N; ++ii){
            for(int jj=0; jj<=k; ++jj)
                dp[ii][jj]= Integer.MAX_VALUE/2;
        }

        //when jj=0 -- we have jj start from 1 -> no reset -> min waste is have sz=mx in the interval 
        int mx=0;
        int ttl=0;
        for(int ii=0; ii<N; ++ii){
            mx=Math.max(mx, nums[ii]);
            ttl += nums[ii];
            dp[ii][0] = mx*(ii+1)-ttl;
        }

        dp[0][0]=0;
        //populate dp
        for(int ii=1; ii<N; ++ii){//for each item in array
            for(int jj=1; jj<=Math.min(ii, k); ++jj){ //for each resize count, cannot reset more than available ii
                mx=0;
                ttl=0;
                for(int st=ii; st>=1; --st){
                    ttl += nums[st];
                    mx=Math.max(mx, nums[st]);
                    dp[ii][jj] = Math.min(dp[ii][jj], dp[st-1][jj-1] + mx*(ii-st+1)-ttl);
                }
            }
        }

        int ret=Integer.MAX_VALUE/2;

        for(int jj=0; jj<=k; ++jj) ret = Math.min(ret, dp[N-1][jj]);

        return ret;
    }
}
